text/0000-variant-types.md

```diff
@@ -0,0 +1,335 @@
+padding to the size of the largest variant. This is to make sharing
+implementations easier (via coercion), see below.
+
+## Conversions
+
+A variant value may be implicitly coerced to its corresponding enum type (an
+upcast). An enum value may be explicitly cast to the type of any of its variants
+(a downcast). Such a cast includes a dynamic check of the discriminant and will
+panic if the cast is to the wrong variant. Variant values may not be converted
+to other variant types. E.g.,
+
+```
+let a: Foo::Variant1 = Foo::Variant1;
+let b: Foo = a; // Ok
+let _: Foo::Variant2 = a; // Compile-time error
+let _: Foo::Variant2 = b; // Compile-time error
+let _ = a as Foo::Variant2; // Compile-time error
+let _ = b as Foo::Variant2; // Runtime error
+let _ = b as Foo::Variant1; // Ok
+```
+
+## impls
+
+`impl`s may exist for both enum and variant types. There is no explicit sharing
+of impls, and just because as enum has a trait bound, does not imply that the
+variant also has that bound. However, the usual conversion rules apply, so if a
+method would apply to the enum type, it can be called on a variant value due to
+coercion performed by the dot operator.
+
+
+# Detailed design - untagged enums
+
+An enum may have `#[repr(union)]` as an attibute. This implies `#[repr(C)]`,
+i.e., variants will have the layout expected for C structs. More importantly, it
+means that the enum is untagged: there is no discriminant. Matching (and `if
+let`, etc.) are not allowed on such enums.
+
+The size of a union value is exactly the size of the largest variant (including
+any padding). There is no discriminant, nor is it possible to have drop flags.
+
+There is no restriction on the kind of variants that can be used with
+`#[repr(union)]`. Unit-like, tuple-like, and struct-like can all be used. Note
+that if all variants are unit-like, then the enum is a zero-sized type. If there
+are other variants, then unit-like variant values are all padding. I don't see
+the utility of such variants, but I see no reason to ban them.
+
+The only operation that can be performed on a union value is casting. An enum
+value can be cast to a variant type. This is not checked (it cannot be, since
+there is no discriminant) and thus is *unsafe*. Variants can also be cast
+'sideways' to other variant types (also unsafe). Like other enums, a variant
+value can be implicitly coerced to the enum type; this is a safe operation.
+
+impls work exactly like regular enums.
+
+## Example
+
+```rust
+#[repr(union)]
+enum MyUnion {
+    MyInt(i64),
+    MyBytes(u8, u8, u8, u8),
+}
+
+fn foo(m: MyUnion) -> i64 {
+    #[import_variant_type]
+    use MyUnion::*;
+
+    assert!(size_of::<MyUnion>() == 8);
+    assert!(size_of::<MyInt>() == 8);
+    assert!(size_of::<MyBytes>() == 8); // 4 bytes of inaccessible padding
+
+    if consult_magic_8_ball() == 42 {
+        unsafe {
+            process_bytes(m as MyBytes)
+        }
+    } else {
+        unsafe { m as MyInt }.0
```

Comment:
A block, including `unsafe {...}` can only be an rvalue which means this copies `MyInt`.

While here it wouldn't make much of a difference, showing `(m as MyInt).0`, maybe even taking a reference thereof or assigning to it, would make the lvalue property clearer.

However, `as` currently only produces rvalues, so I'm not sure what the overall intent is.